mathjax3-ts/core/Tree/Factory.ts

```diff
@@ -0,0 +1,176 @@
+/*************************************************************
+ *
+ *  Copyright (c) 2017 The MathJax Consortium
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/**
+ * @fileoverview  The generic Factory class for creating arbitrary objects
+ *
+ * @author dpvc@mathjax.org (Davide Cervone)
+ */
+
+/*****************************************************************/
+/*
+ * The Factory node interfaces (one for the node instance, one for the node class)
+ */
+
+export interface FactoryNode {
+    readonly kind: string;
+}
+
+export interface FactoryNodeClass<N extends FactoryNode> {
+    new(factory: Factory<N, FactoryNodeClass<N>>, ...args: any[]): N;
+}
+
+/*****************************************************************/
+/*
+ * The Factory interface
+ */
+
+export interface Factory<N extends FactoryNode, C extends FactoryNodeClass<N>> {
+    /*
+     * @param {string} kind  The kind of node to create
+     * @return {N}  The newly created node of the given kind
+     */
+    create(kind: string): N;
+
+    /*
+     * Defines a class for a given node kind
+     *
+     * @param {string} kind  The kind whose class is being defined
+     * @param {C} nodeClass  The class for the given kind
+     */
+    setNodeClass(kind: string, nodeClass: C): void;
+    /*
+     * @param {string} kind  The kind of node whose class is to be returned
+     * @return {C}  The class object for the given kind
+     */
+    getNodeClass(kind: string): C;
+    /*
+     * @param {string} kind  The kind whose definition is to be deleted
+     */
+    deleteNodeClass(kind: string): void;
+
+    /*
+     * @param {N} node  The node to test if it is of a given kind
+     * @param {string} kind  The kind to test for
+     * @return {boolean}  True if the node is of the given kind, false otherwise
+     */
+    nodeIsKind(node: N, kind: string): boolean;
+
+    /*
+     * @return {string[]}  The names of all the available kinds of nodes
+     */
+    getKinds(): string[];
+}
+
+
+/*****************************************************************/
+/*
+ * The generic AbstractFactoryClass interface
+ *   (needed for access to defaultNodes via the constructor)
+ */
+
+interface AbstractFactoryClass<N extends FactoryNode, C extends FactoryNodeClass<N>> extends Function {
```

Comment:
It turns out that the casting in line 123 below needs it.  

```
(this.constructor as AbstractFactoryClass<N, C>)
```

produces the error:

```
Type 'Function' cannot be converted to type 'AbstractFactoryClass<N, C>'.
  Property 'defaultNodes' is missing in type 'Function'.
```

If you know a better way to accomplish this, let me know.